pkg/noun/manage.c

```diff
@@ -2165,7 +2181,51 @@ u3m_migrate(u3v_version ver_w)
 
   /* finally update the version and commit to disk */
   u3H->ver_w = ver_w;
-  /* extra assurance we haven't corrupted the loom before writing to disk */
+  assert(u3m_save_snapshot() == c3y);
+}
+
+c3_o
+u3m_save_snapshot(void)
+{
+  assert(u3_pma);
   u3a_loom_sane();
-  u3e_save();
+
+  if (pma_sync(u3_pma) == -1) {
+    fprintf(stderr, "snapshot: failed to save: %s\r\n", strerror(errno));
+    return c3n;
+  }
+  return c3y;
+}
+
+c3_o
+u3m_backup_snapshot(const c3_c *dir_c)
+{
+  if (!dir_c) {
+    fprintf(stderr, "snapshot: pier directory cannot be NULL\r\n");
+    return c3n;
+  }
+
+  c3_c cmd_c[1024];
+  if (snprintf(cmd_c,
+               sizeof(cmd_c),
+               "mkdir -p %s/.urb/bhk && cp %s/.urb/chk/*.bin %s/.urb/bhk",
+               dir_c,
+               dir_c,
+               dir_c)
+      < 0)
+  {
+    fprintf(stderr,
+            "snapshot: failed to create path needed to backup: %s\r\n",
+            strerror(errno));
+    return c3n;
+  }
+
+  if (system(cmd_c) != 0) {
```

Comment:
This should actually make the appropriate `mkdir()` calls, there is similar code already in this module.

Reply:
Out of curiosity, why is calling `mkdir()` preferred to `system()`?  Both calls to `u3m_backup_snapshot()` occur in locations where performance isn't critical.

Reply:
For the same reason that if you were writing a shell script you'd prefer an explicit `mkdir dir` invocation over `eval "mkdir dir"`. I don't think we should be needlessly invoking the shell if we don't have to. Vastly prefer linking against and directly calling the appropriate system routine.

Reply:
Additionally, it makes debugging difficult since it'll fork the process. Since mars is also forked, we can't just unconditionally set follow-fork-mode to child since we'll step into the shell process. You have to anticipate the system call, switch follow-fork-mode to parent, then switch it back to child.